File: Code/js/Lesson_12_code/src/cacheBench.ts
/**
 * The TypeScript twin of `Code/Lesson_10_code/cache_bench.py`.
 *
 *     PORT=8009 node src/server.ts   # in one terminal
 *     node src/cacheBench.ts         # in another
 *
 * Same three steps: miss against hit, then a stale read, then the fix.
 */

import { Client } from 'pg';
import { createClient } from 'redis';

import { config } from './config.ts';

const BASE_URL = `http://127.0.0.1:${config.port}`;
const QUERY = 'Article number 1370';
const STALE_QUERY = 'Cache demo';

async function timedGet(path: string): Promise<[number, unknown]> {
  const start = performance.now();
  const response = await fetch(`${BASE_URL}${path}`);
  const elapsedMs = performance.now() - start;
  return [elapsedMs, await response.json()];
}

async function benchMissVsHit(redis: ReturnType<typeof createClient>): Promise<void> {
  console.log('--- 1. miss against hit -----------------------------------');
  const key = `search:${QUERY}:0:10`;
  const misses: number[] = [];
  for (let i = 0; i < 5; i++) {
    await redis.del(key);
    const [elapsedMs] = await timedGet(`/bookmarks/search?q=${encodeURIComponent(QUERY)}`);
    misses.push(elapsedMs);
  }

  const hits: number[] = [];
  for (let i = 0; i < 5; i++) {
    const [elapsedMs] = await timedGet(`/bookmarks/search?q=${encodeURIComponent(QUERY)}`);
    hits.push(elapsedMs);
  }

  const avgMiss = misses.reduce((a, b) => a + b, 0) / misses.length;
  const avgHit = hits.reduce((a, b) => a + b, 0) / hits.length;
  console.log('miss (database each time):', misses);
  console.log('hit  (Redis each time):   ', hits);
  console.log(
    `average miss: ${avgMiss.toFixed(3)} ms, average hit: ${avgHit.toFixed(3)} ms, ` +
      `${(avgMiss / avgHit).toFixed(1)}x faster`,
  );
}

async function benchStaleRead(pg: Client): Promise<void> {
  console.log('\n--- 2. the stale read ---------------------------------------');
  await pg.query('DELETE FROM bookmarks WHERE url = $1', ['https://example.com/cache-demo-ts']);

  const [, before] = await timedGet(`/bookmarks/search?q=${encodeURIComponent(STALE_QUERY)}`);
  console.log('before insert:', before);

  await pg.query('INSERT INTO bookmarks (url, title) VALUES ($1, $2)', [
    'https://example.com/cache-demo-ts',
    'Cache demo row ts',
  ]);
  console.log('inserted a matching row directly in PostgreSQL, bypassing the API');

  const [, stillCached] = await timedGet(`/bookmarks/search?q=${encodeURIComponent(STALE_QUERY)}`);
  console.log('right after insert (within the 30s TTL):', stillCached);

  console.log('waiting 31s for the cache entry to expire ...');
  await new Promise((resolve) => setTimeout(resolve, 31_000));

  const [, fresh] = await timedGet(`/bookmarks/search?q=${encodeURIComponent(STALE_QUERY)}`);
  console.log('after the TTL:', fresh);

  await pg.query('DELETE FROM bookmarks WHERE url = $1', ['https://example.com/cache-demo-ts']);
}

async function main(): Promise<void> {
  const redis = createClient({ url: config.redisUrl });
  await redis.connect();
  const pg = new Client({ connectionString: config.databaseUrl });
  await pg.connect();

  try {
    await benchMissVsHit(redis);
    await benchStaleRead(pg);
  } finally {
    await redis.quit();
    await pg.end();
  }
}

main();

File: Code/Lesson_10_code/cache_bench.py
"""What a Redis cache buys you, and what it costs you.

    uvicorn main:app --port 8000    # in one terminal
    python cache_bench.py           # in another

The program does three things:

    1. It clears the cache key, then times a miss and five hits, and reports
       both.
    2. It writes a row directly into PostgreSQL, bypassing the API, and shows
       that a cached search does not see it.
    3. It waits for the cache entry to expire, and shows the same search see
       the new row.

Every number that this program prints comes from your machine. Run it before
you read the lesson section that quotes it.
"""

import time

import httpx
import psycopg
import redis

from config import settings

BASE_URL = "http://127.0.0.1:8000"
QUERY = "Article number 1370"
STALE_QUERY = "Cache demo"


def timed_get(client: httpx.Client, url: str) -> tuple[float, dict]:
    start = time.perf_counter()
    response = client.get(url)
    elapsed_ms = (time.perf_counter() - start) * 1000
    return elapsed_ms, response.json()


def bench_miss_vs_hit(client: httpx.Client, cache: redis.Redis) -> None:
    print("--- 1. miss against hit -----------------------------------")
    key = f"search:{QUERY}:0:10"
    misses = []
    for _ in range(5):
        cache.delete(key)
        elapsed_ms, _ = timed_get(client, f"/bookmarks/search?q={QUERY}")
        misses.append(elapsed_ms)

    hits = []
    for _ in range(5):
        elapsed_ms, _ = timed_get(client, f"/bookmarks/search?q={QUERY}")
        hits.append(elapsed_ms)

    avg_miss = sum(misses) / len(misses)
    avg_hit = sum(hits) / len(hits)
    print(f"miss (database each time): {misses}")
    print(f"hit  (Redis each time):    {hits}")
    print(f"average miss: {avg_miss:.3f} ms, average hit: {avg_hit:.3f} ms, "
          f"{avg_miss / avg_hit:.1f}x faster")


def bench_stale_read(client: httpx.Client) -> None:
    print("\n--- 2. the stale read ---------------------------------------")
    with psycopg.connect(settings.dsn, autocommit=True) as conn:
        conn.execute("DELETE FROM bookmarks WHERE url = %s",
                     ("https://example.com/cache-demo",))

    _, before = timed_get(client, f"/bookmarks/search?q={STALE_QUERY}")
    print(f"before insert: {before}")

    with psycopg.connect(settings.dsn, autocommit=True) as conn:
        conn.execute(
            "INSERT INTO bookmarks (url, title) VALUES (%s, %s)",
            ("https://example.com/cache-demo", "Cache demo row"),
        )
    print("inserted a matching row directly in PostgreSQL, "
          "bypassing the API")

    _, still_cached = timed_get(client, f"/bookmarks/search?q={STALE_QUERY}")
    print(f"right after insert (within the 30s TTL): {still_cached}")

    print("waiting 31s for the cache entry to expire ...")
    time.sleep(31)

    _, fresh = timed_get(client, f"/bookmarks/search?q={STALE_QUERY}")
    print(f"after the TTL: {fresh}")

    with psycopg.connect(settings.dsn, autocommit=True) as conn:
        conn.execute("DELETE FROM bookmarks WHERE url = %s",
                     ("https://example.com/cache-demo",))


def main() -> None:
    cache = redis.Redis.from_url(settings.redis_url, decode_responses=True)
    with httpx.Client(base_url=BASE_URL) as client:
        bench_miss_vs_hit(client, cache)
        bench_stale_read(client)


if __name__ == "__main__":
    main()
